### ISSUE TITLE
Server Side Identify + Page tracking

### ISSUE THREAD

**Issue by axertion** [omgwtfwow/segment-for-wp-by-in8-io#6]
Hi there!

I installed the plugin and was trying to figure out how to get it working with server side events.

I was trying everything, then I finally realized that Identify and Page View tracking is probably not supported for server side event tracking (at least that's my guess from looking at the code)

Is that so?

Is there a specific reason that's the case since the Identify and Page events is probably the most common things used with Segment?

I wanted to ask before I went ahead and just built that into the plugin in case there was a good reason for not doing it.

Thanks!

**Comment by omgwtfwow** [comment]
Hey,

Server side identify calls should be supported if there is a user id [see here](https://github.com/omgwtfwow/segment-for-wp-by-in8-io/blob/645c86d122e31339318451e6bf0ab4f9336e6401/includes/class-segment-for-wp-by-in8-io-track-server.php#L110) as well as further cases in the file.

I think the documentation is unclear on how to use these.

Once you enable an event, you can go to the Identify menu and add 'Associated Events'.

This should add the server side identify calls when those events are fired.

Maybe it isn't working, let me know.

With the Page calls, I didn't support them yet. 

I guess generally I only ever use client side tracking for pages.  I personally combine the server side and client side events into a single 'view' using analytics tools downstream.

But yes, I guess you're right! It's a big use case of WP, after all it is meant for blogs.

I'll turn this into feature request.

**Comment by axertion** [comment]
Server side identify calls are working, but only for the existing events built in of course (not page views).

The main use-case for moving server side page-views is that some Segment Connectors support server side page views, and front-end page loading times tend to get bogged down the more frontend analytic connections you have enabled.

Moving all possible page tracking connections to the server allows me to reduce the front end JS dependencies and get better performance (which is super important now that Google prioritizes it so high).

I spent the entire day going through the plugin code and attempting to make server side page calls work with no success 😥

Is there any guidance you would give if one were to implement it?

My first attempt was to add a `page_view` function to `class-segment-for-wp-by-in8-io-track-server.php` and use the following hook in `class-segment-for-wp-by-in8-io.php` to call it on every page load:
```
$this->loader->add_action('wp', $segment_php, 'page_view', 5, 1);
```

The I updated the `async_task` function to send the page event using:
```
Analytics::page(array(
    "anonymousId" => $ajs_anon_id,
    "category" => $page_category,
    "name" => $page_name,
    "properties" => {
       "url" => $page_url,
       "referrer" => $page_referrer,
    }
));
```

I hackily put this together and it works kind of sporadically. I don't fully understand how the async behavior works because sometimes it queues and fires appropriately and sometimes it doesn't.

Then I just tried dumping the Segment Page call in the `class-segment-for-wp-by-in8-io-page.php` file which works fine every time, but it last the proper async queuing needed for the server side events.

Any help would be greatly appreciated 🙏

**Comment by omgwtfwow** [comment]
Can you try the version of the plugin I just pushed to the Dev branch?

There should be an option in the 'Pages' section to switch on server side tracking.

Unfortunately none of the context seems to be passed automatically by the library, so I'll need to implement it according to the Segment spec later.

[See here](https://github.com/omgwtfwow/segment-for-wp-by-in8-io/commit/f54396263b69a23a7c2e8033d2c1f56dc3170680#diff-5a1243005a08efd8cedacbe1528a2178759cfc625c33b1d8b0c978960d38aadbR698)

**Comment by axertion** [comment]
It's sort of working, but as you said the important properties like `URL`, `Search` `Category` and `Path` are missing.

The biggest issue however is that a single Page view will sent dozens of 404 page events as well:
![image](https://user-images.githubusercontent.com/617549/144640068-7933032f-6919-4ee0-99d5-d14d4a884556.png)

You'll notice when viewing the Pricing page, it also includes dozens of 404s along with it. I believe this is because all the media assets loading for that page are also triggering the Page events.

**Comment by axertion** [comment]
@omgwtfwow - I created a PR for the issues I called out above. I tested locally and it seems to be working now 👍 

[Fix server side Page calls #9](https://github.com/omgwtfwow/segment-for-wp-by-in8-io/pull/9)

![image](https://user-images.githubusercontent.com/617549/144676409-0cc70bbb-13a9-41f4-bf95-92fce2e5ae75.png)

**Comment by omgwtfwow** [comment]
merged this into dev, will test it out and include in next version
thank you!
